server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,523 @@
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.CheckedConsumer;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.env.NodeEnvironment;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.IOError;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.IntPredicate;
+
+public class LucenePersistedStateFactory {
+    private static final Logger logger = LogManager.getLogger(LucenePersistedStateFactory.class);
+    private static final String CURRENT_TERM_KEY = "current_term";
+    private static final String LAST_ACCEPTED_VERSION_KEY = "last_accepted_version";
+    private static final String NODE_ID_KEY = "node_id";
+    private static final String NODE_VERSION_KEY = "node_version";
+    private static final String TYPE_FIELD_NAME = "type";
+    private static final String DATA_FIELD_NAME = "data";
+    private static final String GLOBAL_TYPE_NAME = "global";
+    private static final String INDEX_TYPE_NAME = "index";
+    private static final String INDEX_UUID_FIELD_NAME = "index_uuid";
+
+    private final NodeEnvironment nodeEnvironment;
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    public LucenePersistedStateFactory(NodeEnvironment nodeEnvironment, NamedXContentRegistry namedXContentRegistry) {
+        this.nodeEnvironment = nodeEnvironment;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    CoordinationState.PersistedState loadPersistedState(BiFunction<Long, MetaData, ClusterState> clusterStateFromMetaData)
+        throws IOException {
+
+        final OnDiskState onDiskState = loadBestOnDiskState();
+
+        final List<MetaDataIndex> metaDataIndices = new ArrayList<>();
+        final List<Closeable> closeables = new ArrayList<>();
+        boolean success = false;
+        try {
+            for (final Path path : nodeEnvironment.nodeDataPaths()) {
+                final Directory directory = createDirectory(getMetaDataIndexPath(path, Version.CURRENT.major));
+                closeables.add(directory);
+                final IndexWriterConfig indexWriterConfig = new IndexWriterConfig(new KeywordAnalyzer());
```

Comment:
Do you mean `indexWriterConfig.setRAMBufferSizeMB()`? The default is 16MB AIUI?

Just trying to understand the consequences of that here. If we set a smaller buffer then does that mean we will write more segments, potentially one for each metadata document? I think we want to avoid that. But combined with the `SerialMergeScheduler` we wouldn't be syncing most of those segments, so maybe we need to do both.

Reply:
It means more segments indeed. We wouldn't be syncing most of these segments indeed (though I'm unsure why you make it sound like it has something to do with the merge scheduler?). For the record, stored fields don't go through IndexWriter's buffer but are directly serialized to disk, so the IW RAM buffer really only applies to the indexed fields, which look like they should always be tiny here. So maybe something like 1MB would be good?

Reply:
Thanks, I didn't know this about stored fields, but that's exactly what we want here. Given that the document contents are not buffered I expect even `1MB` is generous enough to result in a single segment per cluster state update. I adjusted this in bb562f8b55f78391b68b901307066801b9565a83.

Reply:
I was expecting the segments that we sync are the results of merging and therefore dependent on the merge scheduler. When investigating this I tried the `NoMergeScheduler` and saw a lot more segments being synced, because of course we were not merging any of them.

However now that I look more closely I now see that both the `SerialMergeScheduler` and the `ConcurrentMergeScheduler` process all pending merges when flushing so in that sense it makes no difference, right?

Reply:
I would need to double check for ConcurrentMergeScheduler, but I can indeed confirm that for SerialMergeScheduler, all the required merges will have been run when flush() returns.